src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs

```diff
@@ -44,112 +44,32 @@ private InnerCache CurrentCache
 
         public ControllerActionInvokerState GetState(ControllerContext controllerContext)
         {
-            // Filter instances from statically defined filter descriptors + from filter providers
-            IFilterMetadata[] filters;
-
             var cache = CurrentCache;
             var actionDescriptor = controllerContext.ActionDescriptor;
 
+            IFilterMetadata[] filters;
             Entry cacheEntry;
-            if (cache.Entries.TryGetValue(actionDescriptor, out cacheEntry))
+            if (!cache.Entries.TryGetValue(actionDescriptor, out cacheEntry))
             {
-                // Deep copy the cached filter items as filter providers could modify them
-                var filterItems = new List<FilterItem>(cacheEntry.FilterItems.Count);
-                for (var i = 0; i < cacheEntry.FilterItems.Count; i++)
-                {
-                    var filterItem = cacheEntry.FilterItems[i];
-                    filterItems.Add(
-                        new FilterItem(filterItem.Descriptor)
-                        {
-                            Filter = filterItem.Filter,
-                            IsReusable = filterItem.IsReusable
-                        });
-                }
+                var filterFactoryResult = FilterFactory.GetCacheableFilters(_filterProviders, controllerContext);
+                filters = filterFactoryResult.Filters;
```

Comment:
Nope. `GetCacheableFilters` includes the result of running the filter providers and cached filters. If you have a suggestion for an alternative name for the method, we could use that here.